mlflow/store/azure_blob_artifact_repo.py

```diff
@@ -0,0 +1,114 @@
+import os
+import re
+
+from six.moves import urllib
+
+from mlflow.entities.file_info import FileInfo
+from mlflow.store.artifact_repo import ArtifactRepository
+from mlflow.utils.file_utils import build_path, get_relative_path, TempDir
+
+
+class AzureBlobArtifactRepository(ArtifactRepository):
+    """
+    Stores artifacts on Azure Blob Storage.
+
+    This repository is used with URIs of the form
+    ``wasbs://<container-name>@<ystorage-account-name>.blob.core.windows.net/<path>``,
+    following the same URI scheme as Hadoop on Azure blob storage. It requires that your Azure
+    storage access key be available in the environment variable ``AZURE_STORAGE_ACCESS_KEY``.
+    """
+
+    def __init__(self, artifact_uri, client=None):
+        if client:
+            self.client = client
+        else:
+            from azure.storage.blob import BlockBlobService
+            (_, account, _) = AzureBlobArtifactRepository.parse_wasbs_uri(artifact_uri)
+            self.client = BlockBlobService(account_name=account,
+                                           account_key=os.environ.get("AZURE_STORAGE_ACCESS_KEY"))
+        super(AzureBlobArtifactRepository, self).__init__(artifact_uri)
+
+    @staticmethod
+    def parse_wasbs_uri(uri):
+        """Parse a wasbs:// URI, returning (container, storage_account, path)."""
+        parsed = urllib.parse.urlparse(uri)
+        if parsed.scheme != "wasbs":
+            raise Exception("Not a WASBS URI: %s" % uri)
+        match = re.match("([^@]+)@([^.]+).blob.core.windows.net", parsed.netloc)
+        if match is None:
+            raise Exception("WASBS URI must be of the form "
+                            "<container>@<account>.blob.core.windows.net")
+        container = match.group(1)
+        storage_account = match.group(2)
+        path = parsed.path
+        if path.startswith('/'):
+            path = path[1:]
```

Comment:
If it doesn't end with a `/`, we just ignore it?

Reply:
This is to check whether it starts with /. The path can also be the empty string if you just wrote wasbs://whatever@whatever.blob.core.windows.net, and in that case we still want to use the whole bucket. We join paths with os.path.join so we should handle a trailing slash correctly.